File: go/helpers/status.go
// SPDX-License-Identifier: MIT

package helpers

import (
	"time"

	vstar "hop.top/vstar"
	"hop.top/vstar/hashing"
)

// statusProp is the wire name for the STATUS property
// (RFC 5545 §3.8.1.11). The RFC scopes a distinct value vocabulary
// to each of VEVENT, VTODO and VJOURNAL; the property name is
// shared.
const statusProp = "STATUS"

// percentProp is the wire name for the VTODO PERCENT-COMPLETE
// property (RFC 5545 §3.8.1.8). Always carries an unsigned-int 0-100.
const percentProp = "PERCENT-COMPLETE"

// validTodoStatus reports whether s is one of the four wire values
// the spec enumerates for VTODO STATUS. Anything else is rejected by
// SetStatus and reported as ok=false by Status.
func validTodoStatus(s vstar.TodoStatus) bool {
	switch s {
	case vstar.TodoNeedsAction, vstar.TodoInProcess, vstar.TodoCompleted, vstar.TodoCancelled:
		return true
	default:
		return false
	}
}

// Status returns the parsed STATUS value of c when present and valid.
// Returns ok=false when STATUS is absent or carries an unrecognized
// value. Status does not enforce that c is a VTODO — callers wanting
// VTODO-specific semantics should gate on c.Type themselves.
func Status(c vstar.Component) (vstar.TodoStatus, bool) {
	p, ok := c.Get(statusProp)
	if !ok {
		return "", false
	}
	s := vstar.TodoStatus(p.Value)
	if !validTodoStatus(s) {
		return "", false
	}
	return s, true
}

// SetStatus writes the STATUS property on c and refreshes the
// X-VSTAR-HASH last. SetStatus is a no-op when:
//
//   - c is nil
//   - c.Type is not CompTodo (the four TodoStatus values are VTODO-
//     specific; VEVENT and VJOURNAL carry their own vocabularies —
//     see SetEventStatus and SetJournalStatus)
//   - s is not one of the four valid TodoStatus constants
//
// The no-op-on-mismatch shape is the Go convention for optional
// mutators on type-mixed inputs: callers don't get a false sense of
// success but also don't have to set up error-handling for a benign
// mistake.
func SetStatus(c *vstar.Component, s vstar.TodoStatus) {
	if c == nil || c.Type != vstar.CompTodo || !validTodoStatus(s) {
		return
	}
	c.Set(vstar.Property{Name: statusProp, Value: string(s)})
	hashing.SetXVSTAR(c)
}

// Complete finalizes a VTODO atomically: STATUS=COMPLETED,
// COMPLETED=t (UTC form), PERCENT-COMPLETE=100, X-VSTAR-HASH
// refreshed last.
//
// No-op when c is nil or c.Type is not CompTodo. The semantic mirrors
// the crm/internal/vcal precedent: if you call Complete you get the
// full set of "done" markers in one call and the stored hash matches.
func Complete(c *vstar.Component, t time.Time) {
	if c == nil || c.Type != vstar.CompTodo {
		return
	}
	c.Set(vstar.Property{Name: statusProp, Value: string(vstar.TodoCompleted)})
	c.SetCOMPLETED(t)
	setPercentCompleteUnchecked(c, percentMax)
	hashing.SetXVSTAR(c)
}

// validEventStatus reports whether s is one of the three wire values
// RFC 5545 §3.8.1.11 enumerates for VEVENT STATUS.
func validEventStatus(s vstar.EventStatus) bool {
	switch s {
	case vstar.EventTentative, vstar.EventConfirmed, vstar.EventCancelled:
		return true
	default:
		return false
	}
}

// validJournalStatus reports whether s is one of the three wire
// values RFC 5545 §3.8.1.11 enumerates for VJOURNAL STATUS.
func validJournalStatus(s vstar.JournalStatus) bool {
	switch s {
	case vstar.JournalDraft, vstar.JournalFinal, vstar.JournalCancelled:
		return true
	default:
		return false
	}
}

// EventStatus returns the parsed STATUS value of c when present and
// valid *for VEVENT*. Returns ok=false when STATUS is absent or
// carries a value outside the VEVENT vocabulary — including values
// that are legal for another component type, so a VTODO's
// NEEDS-ACTION reports ok=false here.
//
// Like Status, EventStatus does not gate on c.Type: it answers
// "does this component carry a VEVENT-shaped STATUS". Callers
// wanting the type check too should test c.Type themselves.
func EventStatus(c vstar.Component) (vstar.EventStatus, bool) {
	p, ok := c.Get(statusProp)
	if !ok {
		return "", false
	}
	s := vstar.EventStatus(p.Value)
	if !validEventStatus(s) {
		return "", false
	}
	return s, true
}

// SetEventStatus writes the STATUS property on c and refreshes the
// X-VSTAR-HASH last. No-op when c is nil, c.Type is not CompEvent,
// or s is not one of the three valid EventStatus constants.
//
// The type guard is what keeps the three STATUS vocabularies from
// bleeding into each other: calling SetEventStatus on a VJOURNAL
// leaves the component untouched.
func SetEventStatus(c *vstar.Component, s vstar.EventStatus) {
	if c == nil || c.Type != vstar.CompEvent || !validEventStatus(s) {
		return
	}
	c.Set(vstar.Property{Name: statusProp, Value: string(s)})
	hashing.SetXVSTAR(c)
}

// JournalStatus returns the parsed STATUS value of c when present
// and valid *for VJOURNAL*. Returns ok=false when STATUS is absent
// or carries a value outside the VJOURNAL vocabulary.
//
// Like Status, JournalStatus does not gate on c.Type.
func JournalStatus(c vstar.Component) (vstar.JournalStatus, bool) {
	p, ok := c.Get(statusProp)
	if !ok {
		return "", false
	}
	s := vstar.JournalStatus(p.Value)
	if !validJournalStatus(s) {
		return "", false
	}
	return s, true
}

// SetJournalStatus writes the STATUS property on c and refreshes
// the X-VSTAR-HASH last. No-op when c is nil, c.Type is not
// CompJournal, or s is not one of the three valid JournalStatus
// constants.
func SetJournalStatus(c *vstar.Component, s vstar.JournalStatus) {
	if c == nil || c.Type != vstar.CompJournal || !validJournalStatus(s) {
		return
	}
	c.Set(vstar.Property{Name: statusProp, Value: string(s)})
	hashing.SetXVSTAR(c)
}

File: rs/src/helpers/status.rs
// SPDX-License-Identifier: MIT

//! The RFC 5545 §3.8.1.11 `STATUS` property.
//!
//! The RFC scopes a distinct value vocabulary to each of VEVENT, VTODO
//! and VJOURNAL while sharing one property name, so there are three
//! getter/setter pairs here and deliberately **no** unifying `Status`
//! type — a port that adds one has diverged.
//!
//! The getters do not gate on the component type: each answers "does
//! this component carry a <TYPE>-shaped STATUS?", so a VTODO's
//! `NEEDS-ACTION` reads as absent through [`event_status`]. The setters
//! do gate, which is what keeps the three vocabularies from bleeding
//! into each other.

use crate::enums::{EventStatus, JournalStatus, TodoStatus};
use crate::hashing;
use crate::model::{Component, Property};
use crate::CompType;
use chrono::{DateTime, Utc};

/// The wire name, shared by all three vocabularies.
const STATUS: &str = "STATUS";

/// The parsed VTODO `STATUS` of `c`, or `None` when the property is
/// absent or carries a value outside the VTODO vocabulary.
///
/// Spelled `todo_status` rather than `status`: in Go the bare name is
/// historical, and every port spells all three pairs symmetrically so
/// no caller reaches for `status` expecting it to work on a VEVENT.
pub fn todo_status(c: &Component) -> Option<TodoStatus> {
    c.get(STATUS).and_then(|p| TodoStatus::parse(&p.value))
}

/// Writes the VTODO `STATUS` and refreshes `X-VSTAR-HASH` last.
///
/// No-op when `c` is not a VTODO — the four [`TodoStatus`] values are
/// VTODO-specific; VEVENT and VJOURNAL carry their own vocabularies.
pub fn set_todo_status(c: &mut Component, s: TodoStatus) {
    if c.r#type != CompType::TODO {
        return;
    }
    c.set(Property::new(STATUS, s.as_str()));
    hashing::set_x_vstar(c);
}

/// The parsed VEVENT `STATUS` of `c`, or `None` when absent or outside
/// the VEVENT vocabulary — including values legal for another component
/// type.
pub fn event_status(c: &Component) -> Option<EventStatus> {
    c.get(STATUS).and_then(|p| EventStatus::parse(&p.value))
}

/// Writes the VEVENT `STATUS` and refreshes `X-VSTAR-HASH` last. No-op
/// when `c` is not a VEVENT.
pub fn set_event_status(c: &mut Component, s: EventStatus) {
    if c.r#type != CompType::EVENT {
        return;
    }
    c.set(Property::new(STATUS, s.as_str()));
    hashing::set_x_vstar(c);
}

/// The parsed VJOURNAL `STATUS` of `c`, or `None` when absent or
/// outside the VJOURNAL vocabulary.
pub fn journal_status(c: &Component) -> Option<JournalStatus> {
    c.get(STATUS).and_then(|p| JournalStatus::parse(&p.value))
}

/// Writes the VJOURNAL `STATUS` and refreshes `X-VSTAR-HASH` last.
/// No-op when `c` is not a VJOURNAL.
pub fn set_journal_status(c: &mut Component, s: JournalStatus) {
    if c.r#type != CompType::JOURNAL {
        return;
    }
    c.set(Property::new(STATUS, s.as_str()));
    hashing::set_x_vstar(c);
}

/// Finalizes a VTODO atomically: `STATUS=COMPLETED`, `COMPLETED=t` in
/// UTC form #2, `PERCENT-COMPLETE=100`, and `X-VSTAR-HASH` refreshed
/// last.
///
/// One call yields the full set of "done" markers with a stored hash
/// that matches; doing it as three separate mutations leaves a window
/// in which the component is half-finished. No-op when `c` is not a
/// VTODO.
pub fn complete(c: &mut Component, t: DateTime<Utc>) {
    if c.r#type != CompType::TODO {
        return;
    }
    c.set(Property::new(STATUS, TodoStatus::Completed.as_str()));
    c.set_completed(t);
    c.set(Property::new(
        super::revision::PERCENT_COMPLETE,
        super::revision::PERCENT_MAX.to_string(),
    ));
    hashing::set_x_vstar(c);
}
